src/index_with_decompressor.ts

```diff
@@ -0,0 +1,120 @@
+import { gameChunks } from './fileReader';
+import { KDRatioMetric, MateAndAssistMetric, KillStreakMetric } from './metrics/captures';
+import { MoveDistanceMetric } from './metrics/distances';
+import { MetadataMetric } from './metrics/misc';
+import {
+  GameWithMostMovesMetric,
+  PieceLevelMoveInfoMetric,
+  MiscMoveFactMetric,
+} from './metrics/moves';
+import { PromotionMetric } from './metrics/promotions';
+import * as fs from 'fs';
+import * as path from 'path';
+import * as async from 'async';
+
+/**
+ *
+ * @param path
+ * @returns
+ */
+export async function main(path: string) {
+  console.time('Total Execution Time');
+  await gameIterator(path);
+  console.timeEnd('Total Execution Time');
+  return results;
+}
+
+let results = {
+  'Number of games analyzed': 0,
+}
+
+/**
+ * Metric functions will ingest a single game at a time
+ * @param metricFunctions
+ */
+async function gameIterator(path) {
+  const cjsmin = new Chess();
+
+  const gamesGenerator = gameChunks(path);
+  const kdRatioMetric = new KDRatioMetric();
+  const killStreakMetric = new KillStreakMetric();
+  const mateAndAssistMetric = new MateAndAssistMetric();
+  const promotionMetric = new PromotionMetric();
+  const moveDistanceMetric = new MoveDistanceMetric();
+  const gameWithMostMovesMetric = new GameWithMostMovesMetric();
+  const pieceLevelMoveInfoMetric = new PieceLevelMoveInfoMetric();
+  const metadataMetric = new MetadataMetric(cjsmin);
+  const miscMoveFactMetric = new MiscMoveFactMetric();
+  const metrics = [
+    metadataMetric,
+    kdRatioMetric,
+    killStreakMetric,
+    mateAndAssistMetric,
+    promotionMetric,
+    moveDistanceMetric,
+    gameWithMostMovesMetric,
+    pieceLevelMoveInfoMetric,
+    miscMoveFactMetric,
+  ];
+
+  let gameCounter = 0;
+  for await (const { moves, metadata } of gamesGenerator) {
+    gameCounter++;
+    if (gameCounter % 400 == 0) {
+      console.log('number of games ingested: ', gameCounter);
+    }
+
+    for (const metric of metrics) {
+      // with array creation
+      const historyGenerator = cjsmin.historyGeneratorArr(moves);
+      metric.processGame(Array.from(historyGenerator), metadata);
+    }
+  }
+  results['Number of games analyzed'] = gameCounter;
+  let metricCallsCount = 0;
+  for (const metric of metrics) {
+    metricCallsCount++;
+    results[metric.constructor.name] = metric.aggregate()
+  }
+}
+
+// Create a write to result.json queue with a concurrency of 1
+const queue = async.queue((task) => {
+  return new Promise<void>((resolve, reject) => {
+    const { results, analysisKey, resultsPath } = task;
+    try {
+      fs.writeFileSync(resultsPath, JSON.stringify(results, null, 2));
+      console.log(`Analysis "${analysisKey}" has been written to ${resultsPath}`);
+      resolve();
+    } catch (err) {
+      reject(err);
+    }
+  });
+}, 1);
+
+// for use with streaming_partial_decompresser.js
+// counter introduced to avoid overwriting existing data in results.json
```

Comment:
what counter?

Reply:
this comment is a relic of a past attempt at avoiding overwriting. removing